From c795d7049b34905c43958d9199d5abcc5862a2e3 Mon Sep 17 00:00:00 2001
From: Seamus Kelly <seamus.kelly@intel.com>
Date: Thu, 20 May 2021 16:42:52 +0100
Subject: [PATCH 098/109] xlink-core: Fix lock issues where a certain code
 paths left channel structure locked and thereby prevent future use of that
 channel

Signed-off-by: Seamus Kelly <seamus.kelly@intel.com>
---
 drivers/misc/xlink-core/xlink-multiplexer.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/misc/xlink-core/xlink-multiplexer.c b/drivers/misc/xlink-core/xlink-multiplexer.c
index b3dca153cba9..c58587597cc2 100644
--- a/drivers/misc/xlink-core/xlink-multiplexer.c
+++ b/drivers/misc/xlink-core/xlink-multiplexer.c
@@ -584,7 +584,7 @@ enum xlink_error xlink_multiplexer_tx(struct xlink_event *event,
 					if (!opchan) {
 						pr_err("link %u, channel %u in invalid state\n",
 						       link_id, chan);
-						return X_LINK_COMMUNICATION_FAIL;
+						rc = X_LINK_COMMUNICATION_FAIL;
 					}
 				} else {
 					rc = X_LINK_CHAN_FULL;
@@ -612,7 +612,7 @@ enum xlink_error xlink_multiplexer_tx(struct xlink_event *event,
 						if (!opchan) {
 							pr_err("link %u, channel %u in invalid state\n",
 							       link_id, chan);
-							return X_LINK_COMMUNICATION_FAIL;
+							rc = X_LINK_COMMUNICATION_FAIL;
 						}
 					}
 				}
@@ -638,7 +638,7 @@ enum xlink_error xlink_multiplexer_tx(struct xlink_event *event,
 				if (!opchan) {
 					pr_err("link %u, channel %u in invalid state\n",
 					       link_id, chan);
-					return X_LINK_COMMUNICATION_FAIL;
+					rc = X_LINK_COMMUNICATION_FAIL;
 				}
 			}
 			if (rc == X_LINK_SUCCESS) {
@@ -677,7 +677,7 @@ enum xlink_error xlink_multiplexer_tx(struct xlink_event *event,
 				if (!opchan) {
 					pr_err("link %u, channel %u in invalid state\n",
 					       link_id, chan);
-					return X_LINK_COMMUNICATION_FAIL;
+					rc = X_LINK_COMMUNICATION_FAIL;
 				}
 			}
 			if (rc == X_LINK_SUCCESS) {
@@ -773,6 +773,7 @@ enum xlink_error xlink_multiplexer_tx(struct xlink_event *event,
 					if (!opchan) {
 						pr_err("link %u, channel %u in invalid state\n",
 						       link_id, chan);
+						mutex_unlock(&xmux->channels[link_id][chan].lock);
 						return X_LINK_COMMUNICATION_FAIL;
 						}
 					if (rc == 0) {
-- 
2.25.1

